File: backend/src/main/java/com/stocktracker/util/PasswordValidator.java
package com.stocktracker.util;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class PasswordValidator {
    private static final int MIN_LENGTH = 8;
    private static final Pattern UPPERCASE = Pattern.compile("[A-Z]");
    private static final Pattern LOWERCASE = Pattern.compile("[a-z]");
    private static final Pattern DIGIT = Pattern.compile("\\d");
    private static final Pattern SYMBOL = Pattern.compile("[!@#$%^&*()_+\\-=\\[\\]{}|;':\",./<>?]");

    public static List<String> validate(String password) {
        List<String> errors = new ArrayList<>();
        
        if (password == null || password.length() < MIN_LENGTH) {
            errors.add("Password must be at least " + MIN_LENGTH + " characters");
        }
        if (password != null) {
            if (!UPPERCASE.matcher(password).find()) {
                errors.add("Password must contain at least one uppercase letter");
            }
            if (!LOWERCASE.matcher(password).find()) {
                errors.add("Password must contain at least one lowercase letter");
            }
            if (!DIGIT.matcher(password).find()) {
                errors.add("Password must contain at least one number");
            }
            if (!SYMBOL.matcher(password).find()) {
                errors.add("Password must contain at least one symbol (!@#$%^&*()_+-=[]{}|;':\",./<>?)");
            }
        }
        return errors;
    }

    public static boolean isValid(String password) {
        return validate(password).isEmpty();
    }
}

File: backend-nodejs/src/common/utils/password-validator.ts
/**
 * Password validation utility matching Java's PasswordValidator
 * Requirements:
 * - Min 8 characters
 * - Max 72 characters (BCrypt limit)
 * - At least one uppercase letter
 * - At least one lowercase letter
 * - At least one digit
 * - At least one special character
 */
export class PasswordValidator {
  private static readonly MIN_LENGTH = 8;
  private static readonly MAX_LENGTH = 72; // BCrypt limit
  private static readonly UPPERCASE = /[A-Z]/;
  private static readonly LOWERCASE = /[a-z]/;
  private static readonly DIGIT = /\d/;
  private static readonly SYMBOL = /[!@#$%^&*()_+\-=[\]{}|;':",./<>?]/;

  /**
   * Validate password and return list of error messages
   * Returns empty array if password is valid
   */
  static validate(password: string | null | undefined): string[] {
    const errors: string[] = [];

    if (!password || password.length < this.MIN_LENGTH) {
      errors.push(`Password must be at least ${this.MIN_LENGTH} characters`);
    }

    if (password && password.length > this.MAX_LENGTH) {
      errors.push(`Password must not exceed ${this.MAX_LENGTH} characters`);
    }

    if (password) {
      if (!this.UPPERCASE.test(password)) {
        errors.push('Password must contain at least one uppercase letter');
      }
      if (!this.LOWERCASE.test(password)) {
        errors.push('Password must contain at least one lowercase letter');
      }
      if (!this.DIGIT.test(password)) {
        errors.push('Password must contain at least one number');
      }
      if (!this.SYMBOL.test(password)) {
        errors.push(
          'Password must contain at least one symbol (!@#$%^&*()_+-=[]{}|;\':",./<>?)',
        );
      }
    }

    return errors;
  }

  /**
   * Check if password is valid (no errors)
   */
  static isValid(password: string | null | undefined): boolean {
    return this.validate(password).length === 0;
  }
}
